amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java

```diff
@@ -0,0 +1,474 @@
+                        log.error("Failed to read entries from exchange {}", exchange.getName(), exception);
+                    }
+                }, null, null);
+            } else {
+                log.warn("{} Not schedule read due to pending read. Messages to read {}.",
+                        exchange.getName(), availablePermits);
+                if (log.isDebugEnabled()) {
+                    log.debug("{} Not schedule read due to pending read. Messages to read {}.",
+                            exchange.getName(), availablePermits);
+                }
+            }
+        } else {
+            // no permits from rate limit
+            exchange.getTopic().getBrokerService().getPulsar().getExecutor()
+                    .schedule(this::readMoreEntries, MESSAGE_RATE_BACKOFF_MS, TimeUnit.MILLISECONDS);
+        }
+    }
+
+    private int getAvailablePermits() {
+        int availablePermits = replicatorQueueSize - PENDING_SIZE_UPDATER.get(this);
+        if (availablePermits <= 0) {
+            log.warn("{} Replicator queue is full, availablePermits: {}, pause route.",
+                    exchange.getName(), availablePermits);
+            if (log.isDebugEnabled()) {
+                log.debug("{} Replicator queue is full, availablePermits: {}, pause route.",
+                        exchange.getName(), availablePermits);
+            }
+            return 0;
+        }
+        return availablePermits;
+    }
+
+    private void routeMessages(List<Entry> entries) throws PulsarServerException {
+        PENDING_SIZE_UPDATER.addAndGet(this, entries.size());
+        for (Entry entry : entries) {
+            Map<String, String> props;
+            MessageImpl<byte[]> message;
+            try {
+                message = MessageImpl.create(null, null,
+                        Commands.parseMessageMetadata(entry.getDataBuffer()),
+                        entry.getDataBuffer(),
+                        Optional.empty(), null, Schema.BYTES, 0, true, -1L);
+                message.getMessageBuilder().clearSequenceId();
+                props = message.getMessageBuilder().getPropertiesList().stream()
+                        .collect(Collectors.toMap(KeyValue::getKey, KeyValue::getValue));
+            } catch (Exception e) {
+                log.error("Deserialize entry dataBuffer failed for exchange {}, skip it first.",
+                        exchange.getName(), e);
+                PENDING_SIZE_UPDATER.decrementAndGet(this);
+                entry.release();
+                continue;
+            }
+
+            Set<Destination> destinations = getDestinations(
+                    props.getOrDefault(MessageConvertUtils.PROP_ROUTING_KEY, ""), getMessageHeaders());
+            initProducerIfNeeded(destinations);
+
+            final Position position = entry.getPosition();
+            List<CompletableFuture<MessageId>> futures = new ArrayList<>();
+            if (!destinations.isEmpty()) {
```

Comment:
if destinations.isEmpty, is it possible to ack exchange topic message?

Reply:
Yes, if the destinations is empty, the futures will be also empty.